### DISCUSSION TITLE
How to best implement a request observer library, that won't clash with Alamofire?

### DISCUSSION THREAD

**Discussion by heitara** [Alamofire/Alamofire#3863] [Q&A]
Hello, I'm trying to understand what's the best way to work side by side with Alamofire. 

The current situation is quite simple.

I have a plain iOS project that uploads a file to a server using Alamofire. The `uploadProgress` callback reports the progress of the upload. So far so good.

Once I add [DebugSwift](https://github.com/heitara/DebugSwift) and the library starts to track the requests the `uploadProgress` callback is not fired. It's similar with Wormholy - here is the exact [issue](https://github.com/pmusolino/Wormholy/issues/77). 

I believe there should be a nice way to get enough information from Alamofire which can be used in those libraries. If Alamofire is not available, then they can work on their own. 

Is there any type of documentation or example that can help me understand what are the best practices to collect information about each request before start reading the code that handles that.

Thanks!

**Comment by jshier** [comment]
Alamofire provides the `EventMonitor` protocol and its various hooks to introspect Alamofire requests. I suggest you read [the documentation](https://github.com/Alamofire/Alamofire/blob/master/Documentation/AdvancedUsage.md#using-eventmonitors) and the [type documentation](https://alamofire.github.io/Alamofire/Protocols/EventMonitor.html). You can hook logging into those events, or into the requests you make directly.
